src/DensityTest.jl

```diff
@@ -0,0 +1,272 @@
+
+# 1. Descriptive Statistics
+
+function DescriptiveStats(RunVar)
+    n = Base.length(RunVar)
+    sd = StatsBase.std(RunVar)
+    rmax = Base.maximum(RunVar)
+    rmin = Base.minimum(RunVar)
+    return n, sd, rmax, rmin
+end
+
+# 2. Binsize (uses DescriptiveStats)
+# function Binsize(n, sd, bin::Union{Real,Nothing}=nothing)
+function Binsize(n, sd)
+    b = 2 * sd * n^(-.5)
+end
+
+# 3. Number of Binsize
+function BinNumber(c, b, rmin, rmax)
+    Jl = convert(Int64, round(floor((c - rmin) / b) + 1))
+    J = convert(Int64, floor((rmax - rmin) / b) + 2)
+    Jr = J - Jl
+    return J, Jl, Jr
+end
+
+# 4. Discretetized running variable (DRV).
+function DRV(RunVar, c, b, rmin, J, n)
+    l  = floor((rmin - c) / b) * b + 0.5 * b + c
+
+    # Discretized running variable 
+    X = zeros(J)
+    for j = 1:J
+        X[j] = l + (j - 1) * b
+    end
+
+    gR = floor.((RunVar .- c) ./ b) * b .+ 0.5 * b .+ c;
+
+    # Creating the smooth histogram Yi
+    Y = zeros(J)
+    for j = 1:J
+        Y[j] = sum(1.0 * (gR .≈ X[j])) / (n * b)
+    end
+    return X, Y
+end
+
+# 5. Bandwidth, needs (4), (3), (2), (1)
+
+function McCraryBandwidth(X, Y, Jl, c)
+        # I will proceed to estimate the optimal bandwidth using McCrary (2008) optimal method.
+    κ = 3.348
+            
+        # Left part
+    ldf = DataFrame(Y=Y[1:Jl], X=X[1:Jl])
+    lols = lm(@formula(Y ~ X + X^2 + X^3 + X^4 ), ldf)
+    lcoef = coef(lols)
+    lmse =  deviance(lols) / dof_residual(lols)
+    lf = 2 * lcoef[3] .+ 6 * lcoef[4] * ldf.X + 12 * lcoef[5] * ldf.X.^2 
+            
+    lh = κ * (lmse * (c - ldf.X[1]) / sum(lf'lf))^(0.2)
+            
+        # Right part
+    rdf = DataFrame(Y=Y[Jl + 1:end], X=X[Jl + 1:end])
+    rols = lm(@formula(Y ~ X + X^2 + X^3 + X^4 ), rdf)
+    rcoef = coef(rols)
+    rmse =  deviance(rols) / dof_residual(rols)
+            
+            
+    rf = 2 * rcoef[3] .+ 6 * rcoef[4] * rdf.X + 12 * rcoef[5] * rdf.X.^2 
+    rh = κ * (rmse * (rdf.X[end] - c) / sum(rf'rf))^(0.2)
+            
+        # Taking the average
+    h = 0.5 * (rh + lh)
+end
+
+# 6. McCraryTest
+"""
+    fit(method::McCraryTest, data::RunningVariable)
+
+Test for manipulation in the running variable following McCrary (2008).
+
+# Arguments
+
+- `data::RunningVariable`: The running variable.
+
+# Returns
+Returns  a FittedMcCraryTest object with fields `b` and `h` which are the bin size and the bandwidth used, respectively. Also, it returns `θhat`, `σθ`, `pval`, `Js`, and `fitted_res` which are the estimated difference in the intercepts, the standard error of the estimate, the pval of the test H0: θhat = 0, the number of bins used, and the fitted values for the support of the running variable (as a DataFrame). Finally, `coeftable` provides an object of type CoefTable for the estimator. 
+"""
+function fit(method::McCraryTest, data::RunningVariable)
+    
+    # Cutoff selection
+    c = data.cutoff
+    R = data.Zs
+    # Call the functions 
```

Comment:
Note that `data.ZsC` stores the centered running variable which means you do not need to be centering later on (X-c) etc since `c` has already been subtracted.

Reply:
I will try to use this! However, I did not include it in this version.